### PR TITLE
feat(DMS): add datasource rocketmq topics

### PR DESCRIPTION
**Pull request by 346010762**
**What this PR does / why we need it**:

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
   add datasource rocketmq topics
```

## PR Checklist

* [ ] Tests added/passed.
* [ ] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST=./huaweicloud/services/acceptance/dms/ TESTARGS='-run TestAccDmsRocketMQTopics_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dms/ -v -run TestAccDmsRocketMQTopics_basic -timeout 360m -parallel 4
=== RUN   TestAccDmsRocketMQTopics_basic
=== PAUSE TestAccDmsRocketMQTopics_basic
=== CONT  TestAccDmsRocketMQTopics_basic
--- PASS: TestAccDmsRocketMQTopics_basic (715.79s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dms       715.827s


```

### CODE REVIEW

File: docs/data-sources/dms_rocketmq_topics.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+---
+subcategory: "Distributed Message Service (DMS)"
+---
+
+# huaweicloud_dms_rocketmq_topics
+
+Use this data source to get the list of DMS rocketMQ topic.
```
Comment (houpeng80) [inline docs/data-sources/dms_rocketmq_topics.md]:
```suggestion
Use this data source to get the list of DMS rocketMQ topics.
```

File: docs/data-sources/dms_rocketmq_topics.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+```hcl
+var "instance_id" {}
+
+data "huaweicloud_dms_rocketmq_topics" "test" {
+  instance_id = var.instance_id
+  name                  = "topic1"
+  total_read_queue_num  = 3
+  total_write_queue_num = 3
+  permission            = "all"
+}
```
Comment (houpeng80) [inline docs/data-sources/dms_rocketmq_topics.md]: align the equal signs

File: docs/data-sources/dms_rocketmq_topics.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+* `instance_id` - (Required, String) Specifies the ID of the rocketMQ instance.
+
+* `name` - (Optional, String) Specifies the topic name.
+
+* `total_read_queue_num` - (Optional, Int) Specifies the number of total read queue.
+
+* `total_write_queue_num` - (Optional, Int) Specifies the number of total write queue.
+
+* `permission` - (Optional, String) Specifies the permission, the possible value is sub, pub or all.
```
Comment (houpeng80) [inline docs/data-sources/dms_rocketmq_topics.md]:
```suggestion
* `permission` - (Optional, String) Specifies the permission. Value options: **sub**, **pub** or **all**.
```

File: docs/data-sources/dms_rocketmq_topics.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+* `message_type` (Optional, String) Specifies the message_type, which will be supported in RocketMQ instance V5.x.
+  The possible value is NORMAL, FIFO, DELAY or TRANSACTION
```
Comment (houpeng80) [inline docs/data-sources/dms_rocketmq_topics.md]: delete it

File: docs/data-sources/dms_rocketmq_topics.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+## Attribute Reference
+
+In addition to all arguments above, the following attributes are exported:
+
```
Comment (houpeng80) [inline docs/data-sources/dms_rocketmq_topics.md:43]: the id is missing

File: docs/data-sources/dms_rocketmq_topics.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+* `topics` - Indicates the list of the topic.
+
+The `topics` block contains:
+
+* `name` - Specifies the topic name.
+
+* `total_read_queue_num` - (Int) Specifies the number of total read queue.
```
Comment (houpeng80) [inline docs/data-sources/dms_rocketmq_topics.md]:
```suggestion
* `total_read_queue_num` - Indicates the number of total read queue.
```
fix other places

File: docs/data-sources/dms_rocketmq_topics.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+* `total_write_queue_num` - (Int) Specifies the number of total write queue.
+
+* `permission` - (String) Specifies the permission, the possible value is sub, pub or all.
+
+* `message_type` (String) Specifies the message_type, which will be supported in RocketMQ instance V5.x.
+  The possible value is NORMAL, FIFO, DELAY or TRANSACTION.
```
Comment (houpeng80) [inline docs/data-sources/dms_rocketmq_topics.md]: delete it

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+	region := cfg.GetRegion(d)
+
+	// getRocketmqTopics: Query the List of rocketMQ topic
+	var (
+		getRocketmqTopicHttpUrl = "v2/{project_id}/instances/{instance_id}/topics"
+		getRocketmqTopicProduct = "dmsv2"
+	)
+	getRocketmqTopicClient, err := cfg.NewServiceClient(getRocketmqTopicProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating DMS rocketmq topics Client: %s", err)
```
Comment (houpeng80) [inline huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go]:
```suggestion
		return diag.Errorf("error creating DMS Client: %s", err)
```

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+	result := make([]interface{}, 0, len(topicArray))
+
+	paramName, statusName := d.GetOk("name")
+	paramTotalReadQueueNum, statusTotalReadQueueNum := d.GetOk("total_read_queue_num")
+	paramTotalWriteQueueNum, statusTotalWriteQueueNum := d.GetOk("total_write_queue_num")
+	paramPermission, statusPermission := d.GetOk("permission")
+	for _, topic := range topicArray {
+		name := utils.PathSearch("name", topic, nil)
+		totalReadQueueNum, _ := utils.PathSearch("total_read_queue_num", topic, nil).(float64)
+		totalWriteQueueNum, _ := utils.PathSearch("total_write_queue_num", topic, nil).(float64)
```
Comment (houpeng80) [inline huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go]:
```suggestion
		totalReadQueueNum, _ := utils.PathSearch("total_read_queue_num", topic, 0).(float64)
		totalWriteQueueNum, _ := utils.PathSearch("total_write_queue_num", topic, 0).(float64)
```

File: docs/data-sources/dms_rocketmq_topics.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  The possible value is NORMAL, FIFO, DELAY or TRANSACTION
+
+## Attribute Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `topics` - Lists the topics.
+  The [topics](#dms_rockermq_topics) structure is documented below.
+
+<a name="dms_rockermq_topics"></a>
```
Comment (houpeng80) [inline docs/data-sources/dms_rocketmq_topics.md]:
```suggestion
  The [topics](#DMS_rockermq_topics) structure is documented below.

<a name="DMS_rockermq_topics"></a>
```

File: docs/data-sources/dms_rocketmq_topics.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+* `permission` - (String) Specifies the permission, the possible value is sub, pub or all.
+
+* `message_type` (String) Specifies the message_type, which will be supported in RocketMQ instance V5.x.
+  The possible value is NORMAL, FIFO, DELAY or TRANSACTION.
+
+* `brokers` - Lists the brokers.
+  The [brokers](#dms_rockermq_topic_brokers) structure is documented below.
+
+<a name="dms_rockermq_topic_brokers"></a>
+The `brokers` block supports:
```
Comment (houpeng80) [inline docs/data-sources/dms_rocketmq_topics.md]:
```suggestion
  The [brokers](#DMS_rockermq_topic_brokers) structure is documented below.

<a name="DMS_rockermq_topic_brokers"></a>
The `brokers` block supports:
```

File: huaweicloud/services/acceptance/dms/data_source_huaweicloud_dms_rocketmq_topics_test.go [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+	})
+}
+
+func testAccDmsRocketMQSearchTopic(name string) string {
+	return fmt.Sprintf(`
+%[1]s
+
+data "huaweicloud_dms_rocketmq_topics" "test" {
+  depends_on            = [huaweicloud_dms_rocketmq_topic.test]
+  instance_id           = huaweicloud_dms_rocketmq_instance.test.id
```
Comment (houpeng80) [inline huaweicloud/services/acceptance/dms/data_source_huaweicloud_dms_rocketmq_topics_test.go]: delete the excess spaces, only one is needed

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+	getRocketmqTopicsPath = strings.ReplaceAll(getRocketmqTopicsPath, "{project_id}", getRocketmqTopicsClient.ProjectID)
+	getRocketmqTopicsPath = strings.ReplaceAll(getRocketmqTopicsPath, "{instance_id}", fmt.Sprintf("%v", d.Get("instance_id")))
+
+	listTopicsResp, err := pagination.ListAllItems(
+		getRocketmqTopicsClient,
+		"offset",
+		getRocketmqTopicsPath,
+		&pagination.QueryOpts{MarkerField: ""})
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "error retrieving DMS rocketmq topics")
```
Comment (houpeng80) [inline huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go]:
```suggestion
		return common.CheckDeletedDiag(d, err, "error retrieving DMS rocketMQ topics")
```

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+	dataSourceId, err := uuid.GenerateUUID()
+	if err != nil {
+		return diag.Errorf("unable to generate ID: %s", err)
+	}
+	d.SetId(dataSourceId)
+
+	mErr := multierror.Append(
+		d.Set("region", region),
+		d.Set("topics", flattenListTopicsBody(filteredTopics)),
+	)
```
Comment (houpeng80) [inline huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go]:
```suggestion

	dataSourceId, err := uuid.GenerateUUID()
	if err != nil {
		return diag.Errorf("unable to generate ID: %s", err)
	}
	d.SetId(dataSourceId)

	mErr := multierror.Append(
		d.Set("region", region),
		d.Set("topics", flattenListTopicsBody(filterTopics(d, listTopicsRespBody))),
	)
```

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+	}
+	rst := make([]interface{}, 0, len(resp))
+
+	for _, v := range resp {
+		rst = append(rst, map[string]interface{}{
+			"name":                  utils.PathSearch("name", v, nil),
+			"total_read_queue_num":  utils.PathSearch("total_read_queue_num", v, nil),
+			"total_write_queue_num": utils.PathSearch("total_write_queue_num", v, nil),
+			"permission":            utils.PathSearch("permission", v, nil),
+			"brokers":               utils.PathSearch("brokers", v, make([]interface{}, 0)),
```
Comment (houpeng80) [inline huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go]:
```suggestion
			"brokers":               utils.PathSearch("brokers", v, nil),
```

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+	topicJson := utils.PathSearch("topics", resp, make([]interface{}, 0))
+	topicArray := topicJson.([]interface{})
+	if len(topicArray) < 1 {
+		return nil
+	}
+	result := make([]interface{}, 0, len(topicArray))
+
+	paramName, statusName := d.GetOk("name")
+	paramTotalReadQueueNum, statusTotalReadQueueNum := d.GetOk("total_read_queue_num")
+	paramTotalWriteQueueNum, statusTotalWriteQueueNum := d.GetOk("total_write_queue_num")
```
Comment (houpeng80) [inline huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go]:
```suggestion
	rawName, rawNameOk:= d.GetOk("name")
	rawTotalReadQueueNum, rawTotalReadQueueNumOk := d.GetOk("total_read_queue_num")
	rawTotalWriteQueueNum, rawTotalWriteQueueNumOk := d.GetOk("total_write_queue_num")
```

File: docs/data-sources/dms_rocketmq_topics.md (hunk 1 above)
Comment (346010762) [reply docs/data-sources/dms_rocketmq_topics.md]: done

File: docs/data-sources/dms_rocketmq_topics.md (hunk 2 above)
Comment (346010762) [reply docs/data-sources/dms_rocketmq_topics.md]: done

File: docs/data-sources/dms_rocketmq_topics.md (hunk 3 above)
Comment (346010762) [reply docs/data-sources/dms_rocketmq_topics.md]: done

File: docs/data-sources/dms_rocketmq_topics.md (hunk 4 above)
Comment (346010762) [reply docs/data-sources/dms_rocketmq_topics.md]: done

File: docs/data-sources/dms_rocketmq_topics.md (hunk 5 above)
Comment (346010762) [reply docs/data-sources/dms_rocketmq_topics.md:43]: done

File: docs/data-sources/dms_rocketmq_topics.md (hunk 6 above)
Comment (346010762) [reply docs/data-sources/dms_rocketmq_topics.md]: done

File: docs/data-sources/dms_rocketmq_topics.md (hunk 7 above)
Comment (346010762) [reply docs/data-sources/dms_rocketmq_topics.md]: done

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go (hunk 8 above)
Comment (346010762) [reply huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go]: done

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go (hunk 9 above)
Comment (346010762) [reply huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go]: done

File: docs/data-sources/dms_rocketmq_topics.md (hunk 10 above)
Comment (346010762) [reply docs/data-sources/dms_rocketmq_topics.md]: done

File: docs/data-sources/dms_rocketmq_topics.md (hunk 11 above)
Comment (346010762) [reply docs/data-sources/dms_rocketmq_topics.md]: done

File: huaweicloud/services/acceptance/dms/data_source_huaweicloud_dms_rocketmq_topics_test.go (hunk 12 above)
Comment (346010762) [reply huaweicloud/services/acceptance/dms/data_source_huaweicloud_dms_rocketmq_topics_test.go]: done

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go (hunk 13 above)
Comment (346010762) [reply huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go]: done

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go (hunk 14 above)
Comment (346010762) [reply huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go]: done

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go (hunk 15 above)
Comment (346010762) [reply huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go]: done

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go (hunk 16 above)
Comment (346010762) [reply huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go]: done

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+)
+
+func DataSourceDmsRocketMQTopics() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: resourceDmsRocketMQTopicsRead,
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (ShiChangkuo) [inline huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go:27]: it's better to add `Computed` attr

File: docs/data-sources/dms_rocketmq_topics.md [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+
+* `permission` - (Optional, String) Specifies the permission. Value options: **sub**, **pub** or **all**.
+
+## Attribute Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The data source ID.
+
+* `topics` - Lists the topics.
```
Comment (ShiChangkuo) [inline docs/data-sources/dms_rocketmq_topics.md]:
```suggestion
* `topics` - The list of topics.
```

File: huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go (hunk 17 above)
Comment (346010762) [reply huaweicloud/services/dms/data_source_huaweicloud_dms_rocketmq_topics.go:27]: done

File: docs/data-sources/dms_rocketmq_topics.md (hunk 18 above)
Comment (346010762) [reply docs/data-sources/dms_rocketmq_topics.md]: done

**Pull request merged**